torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,344 @@
+        # Save the bucket reference and all-reduce future for the final bucket
+        if rank_to_update == rank:
+            overlap_info.bucket_index_to_bucket[bucket_index] = bucket
+            overlap_info.bucket_index_to_future[bucket_index] = fut
+
+        num_buckets = len(overlap_info.params_per_bucket)
+        # NOTE: The implementation from this point forward assumes that the
+        # buckets are indexed incrementally starting from 0 in the order of
+        # their autograd hooks firing
+        is_last_bucket = bucket_index == num_buckets - 1
+        if not is_last_bucket:
+            return fut
+
+        # Perform partial optimizer step on all buckets after the final
+        # bucket has been computed
+        for bucket_index in range(num_buckets):
+            rank_to_update = zero._ddp_bucket_index_to_rank(bucket_index)
+            num_local_optim_params = len(zero.optim.param_groups[0]["params"])
+            if rank_to_update == rank:
+                gradients: List[Optional[torch.Tensor]] = \
+                    [_NO_PARAM_UPDATE for _ in range(num_local_optim_params)]
+                assert bucket_index in overlap_info.offsets, \
+                    f"Bucket index {bucket_index} was not assigned to rank {rank}"
+                offset = overlap_info.offsets[bucket_index]
+                # Ensure that the all-reduce completes before performing the
+                # the parameter update
+                assert bucket_index in overlap_info.bucket_index_to_future, \
+                    f"All-reduce future for bucket {bucket_index} not saved " \
+                    f"on rank {rank}"
+                allreduce_future = overlap_info.bucket_index_to_future[bucket_index]
+                allreduce_future.wait()
+                bucket_gradients = overlap_info.bucket_index_to_bucket[bucket_index].get_per_parameter_tensors()
+                for i, grad in enumerate(bucket_gradients):
+                    gradients[offset + i] = grad
+                zero._local_step(gradients)
+            device = overlap_info.params_per_bucket[bucket_index][0].device
+            device_index = zero._device_to_device_index[device]
+            assert bucket_index in zero._buckets[device_index][rank_to_update]
+            overlap_info.broadcast_handles.append(
+                dist.broadcast(
+                    zero._buckets[device_index][rank_to_update][bucket_index],
+                    src=rank_to_update,
+                    async_op=True
+                )
+            )
+
+        # Ensure that all parameter updates are finished before the
+        # next forward pass
+        assert len(overlap_info.broadcast_handles) == num_buckets, \
+            f"Missing at least one broadcast handle on rank {rank}"
+        _ = list(map(lambda x: x.wait(), overlap_info.broadcast_handles))
+        overlap_info.broadcast_handles.clear()
+
+        # Reset per-iteration information
+        overlap_info.bucket_index_to_future.clear()
```

Comment:
nit (can be done in followup PR, and this is just my personal preference, please feel free to ignore):  The `_OverlapInfo ` type is not owned by this hook function and lives a bit far away from this file. It might be better to wrap its state manipulations into member method in `_OverlapInfo`, instead of directly manipulating them here and above.

Reply:
I think this definitely makes sense. I will try to refactor this in a follow-up PR.